Documentation/design-docs/PinnedHeap.md

```diff
@@ -0,0 +1,160 @@
+
+With these efforts we were able to mitigate perf problems caused by pinned by a lot. But at the end of the day, due to the nature of “pinning already allocated objects”, we cannot eliminate the situation of interleaved pinned and non pinned objects and that can still cause perf problems for GC. And the long lived scattered pins are the most problematic. They can be separated into 2 categories –
+
+<span style="color:blue">**Long lived scattered pins in young generations, especially gen0.**</span>
+
+However, For example, we could have a bunch of long lived pins that get demoted to gen0 with lots of free spaces in between them. And we may not want to reuse all the free spaces (ie, fill all the gen0 free spaces before triggering a GC) because we wouldn’t want to make gen0 pause times too high. We have seen many profiles where the gen0 size is quite significant with a very high fragmentation ratio due to pinning. 
+
+<span style="color:blue">**Long lived scattered pins in old generations**</span>
+
+Nowadays we aim to do background GCs most of the time. And since background GCs do not compact, pinning or not makes no difference. But when we do want to compact gen2 this becomes a problem. 
+
+We do have plans to make the heap into regions instead of segments which would mitigate this situation because we can repurpose the regions with pins for other generations if needed but while it’s easy to repurpose a region with pins that belongs to an ephemeral generation as a gen2 region, the other way is not so easy because we’d have to care about setting cards. And since we know that these are long lived pins, it actually wasn’t desirable to even demote them to gen0 in the first place – we just did that so we could reuse the free spaces between them sooner.
+
+Another issue, even with regions, is that of course the pins will always mean we’d have to keep the whole region there instead of just release the memory of the region back to the OS (sure we could decommit the pages without the pins in the region – obviously that means more bookkeeping). Having regions that are mostly empty and only occupied by a few pins means we delay releasing memory of these regions.
+
+## Solution
+
+The problem with pinned objects is that even a small number of them can have a significant impact on perf due to their positions relative to non pinned objects. And we do make the assumption that there are way fewer pinned objects than non pinned ones. So it makes sense to group pinned ones in their own group so they do not interleave with non pinned ones and we call this group the Pinned Heap, like how we group small and large objects together, ie, SOH (Small Object Heap) and LOH (Large Object Heap). 
+
+In order to group pinned objects together we have a couple of options –
+
+1) We allow pinning at allocation time so when such an object is allocated we put it on the pinned heap. 
+
+2) We allow migrating non pinned objects to the pinned heap at the time they need to be pinned.
```

Comment:
Would this be generally better, ignoring implementation cost? Its nice if it can make existing code better without requiring modifications.

Reply:
I doubt that it would be generally better. This option would change the performance characteristics of pinning in dramatic way. Migrating the object to pinned heap would involve copying and/or adding extra indirections. Everybody would pay for this, even the places that do not benefit.